---
title: "Introduction_DeregGenes"
date: "`r format(Sys.time(), '%d %b %Y')`"
output: 
  rmarkdown::html_vignette:
    toc: true
    number_sections: false
vignette: >
  %\VignetteIndexEntry{Introduction_DeregGenes}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  message = FALSE,
  warning = FALSE,
  comment = "#>"
)
```

```{r setup}
library(DeregGenes)

# load in helper packages for the demo
library(GEOquery)
```

## Introduction

The main objective of **DeregGenes** is to find genes that are deregulated(up-regulate and down- regulate) in different diseases. This package could also allow users to put different results together to generate a heatmap for cross studies analysis by function __*prepareData*__ . It improves users’ time on massive data cleaning and data annotating processes prior to the analysis since different data prepared by different platforms will need to be handled by different tools. This package also allows user to compute gene differential expression data for single study using function __*logFCsingle*__. Moreover, the function __*Aggreg*__ provides a simple way to let users get a summarised result across multiple current studies data to provide them with a more confident result and conclusion. Also, using this package will save users time to switch back and forth between different distinct packages and learn different documentation since the different packages will require different input data types which is not available on any current published r package. Lastly, the package can generate heat maps of gene differential expression across stuides using function __*plotHeatMap*__.

See `help(package = "DeregGenes")` for further details and references provided by `citation("DeregGenes")`. To download **DeregGenes**, use the following commands:

``` r
require("devtools")
install_github("wezhubb/DeregGenes", build_vignettes = TRUE)
library("DeregGenes")
```

To list all functions available in the package:
``` r
lsf.str("package:DeregGenes")
```

## Workflow
In the following session, I will demonstated the use of **DeregGenes** by analysis of gene expression data from GEO GSE84402 and GSE29721. 

### step 1: preparing and annotating data 
Before the formal analysis, it is important to clean the data by removing NULL value, and annotate data by convert the porbe ID/gene ID to universal HGNC gene symbols. The reason for preparing and annotating data is to make any further analysis easier to perform.

To download the data in current directory:

```{r download_data}
# download data1(GSE29721) from GEO into current directory
filePaths1 <- getGEOSuppFiles("GSE29721")

# download data2 from GEO from GEO into current directory
filePaths2 <- getGEOSuppFiles("GSE84402")

```

Because the downloaded file is ziped, we need to unzip the file and delete the orignal zip file:

```{r zip_data}
# untar downloaded data1 and delete tar file
untarPath1 <- strsplit(row.names(filePaths1), '/')
untarPath1 <- paste(untarPath1[[1]][1:length(untarPath1[[1]]) - 1],
     collapse="/")
untar(row.names(filePaths1), exdir = untarPath1)
unlink(paste(untarPath1, '/*.tar', sep = ''))

# untar downloaded data2 and delete tar file
untarPath2 <- strsplit(row.names(filePaths2), '/')
untarPath2 <- paste(untarPath2[[1]][1:length(untarPath2[[1]]) - 1],
     collapse="/")
untar(row.names(filePaths2), exdir = untarPath2)
unlink(paste(untarPath2, '/*.tar', sep = ''))

```

Prepare data:

```{r prepare_data}
# preparing data1
data1 <- prepareData(untarPath1, TRUE)
head(data1)

# preparing data2
data2 <- prepareData(untarPath2, TRUE)
head(data2)

```


After finishing the preparing process and stored the data, we can delete the download files:

```{r delete_data}
# delete all download data1 from current directory
unlink(untarPath1, recursive = TRUE)

# delete all download data2 from current directory
unlink(untarPath2, recursive = TRUE)
```

### step 2: compute differential expression data for each study
With the data being cleaned and annotated, we can now compute the gene differnetial expression/expressional fold change.

Compute differential expression for GSE29721

```{r logFC_data1}
# class is the order of the each sample type in GSE29721
class <- c("mutant", "control","mutant", "control","mutant", "control",
     "mutant", "control","mutant", "control","mutant", "control", "mutant",
     "control","mutant", "control","mutant", "control", "mutant", "control")
result1 <- logFCsingle(data1, class)

head(result1)

```

Compute differential expression for GSE84402

```{r logFC_data2}
# class is the order of the each sample type in GSE84402
class <- c("mutant", "control","mutant", "control","mutant", "control",
     "mutant", "control","mutant", "control","mutant", "control", "mutant",
     "control","mutant", "control","mutant", "control", "mutant", "control",
     "mutant", "control","mutant", "control", "mutant", "control","mutant",
     "control")
result2 <- logFCsingle(data2, class)

head(result2)
```

### step 3: Aggregate differential expression from GSE29721 and GSE84402

With the single study differential expression data with fold change, we need to aggregate them together to find each genes overall performance across all studies.

```{r aggregate}
# create list of input  data
listLogFC <- list(result1, result2)

# create list of data title
listTitle <- c("GSE29721", "GSE84402")

# aggregate
aggreg <- Aggreg(listLogFC, listTitle)

head(aggreg[1])
head(aggreg[2])
head(aggreg[3])

```

### step 4: drawing heat map
With the aggregate data, we can draw the heatmap now.

Notice that the comment below will not be execute here since the purpose of the following function is to store the generated heat map in your local directory and return NULL. The heatmap for the previous example generated by this function is attached below. 

```{r heatmap}
# plotHeatMap(data.frame(aggreg[3]), 4, 6)

```
![](../inst/extdata/logFC.jpg) { width=100% }


## References
* Carvalho B. S., and Irizarry, R. A. 2010. A Framework for Oligonucleotide Microarray Preprocessing Bioinformatics.

* Gautier, L., Cope, L., Bolstad, B. M., and Irizarry, R. A. 2004. affy---analysis of Affymetrix GeneChip data at the probe level. Bioinformatics 20, 3 (Feb. 2004), 307-315.

* Mapping identifiers for the integration of genomic datasets with the R/Bioconductor package biomaRt. Steffen Durinck, Paul T. Spellman, Ewan Birney and Wolfgang Huber, Nature Protocols 4, 1184-1191 (2009).

* BioMart and Bioconductor: a powerful link between biological databases and microarray data analysis. Steffen Durinck, Yves Moreau, Arek Kasprzyk, Sean Davis, Bart De Moor, Alvis Brazma and Wolfgang Huber, Bioinformatics 21, 3439-3440 (2005).

* Wickham H, Averick M, Bryan J, Chang W, McGowan LD, François R, Grolemund G, Hayes A, Henry L, Hester J, Kuhn M, Pedersen TL, Miller E, Bache SM, Müller K, Ooms J, Robinson D, Seidel DP, Spinu V, Takahashi K, Vaughan D, Wilke C, Woo K, Yutani H (2019). “Welcome to the tidyverse.” _Journal of Open Source Software_, *4*(43), 1686. doi:10.21105/joss.01686 <https://doi.org/10.21105/joss.01686>.

* Wickham H, François R, Henry L, Müller K (2022). _dplyr: A Grammar of Data Manipulation_. R package version 1.0.10, <https://CRAN.R-project.org/package=dplyr>.

* Ritchie, M.E., Phipson, B., Wu, D., Hu, Y., Law, C.W., Shi, W., and Smyth, G.K. (2015). limma powers differential expression analyses for RNA-sequencing and microarray studies. Nucleic Acids Research 43(7), e47.

* Hastie T, Tibshirani R, Narasimhan B, Chu G (2022). _impute: impute: Imputation for microarray data_. R package version 1.70.0.

* Kolde R (2022). _RobustRankAggreg: Methods for Robust Rank Aggregation_. R package version 1.2.1, <https://CRAN.R-project.org/package=RobustRankAggreg>.

* Kolde R (2019). _pheatmap: Pretty Heatmaps_. R package version 1.0.12, <https://CRAN.R-project.org/package=pheatmap>.

* Wang H, Huo X, Yang XR, He J et al. STAT3-mediated upregulation of lncRNA HOXD-AS1 as a ceRNA facilitates liver cancer metastasis by regulating SOX4. Mol Cancer 2017 Aug 14;16(1):136. PMID: 28810927

* Stefanska B, Huang J, Bhattacharyya B, Suderman M et al. Definition of the landscape of promoter DNA hypomethylation in liver cancer. Cancer Res 2011 Sep 1;71(17):5891-903. PMID: 21747116

```{r}
sessionInfo()
```
